# Tarea 3.2: Ejercicios con Dataframes.

Introducción a la librería Pandas, a los Dataframes y los métodos. Análisis exploratorio de los datos.

In [1]:
import pandas as pd
import numpy as np
import math

## Ejercicio 1

Descarga el Dataset de Airlines Delay: Airline on-time statistics and delay causas:
* Cárgalo en un Pandas Dataframe. 
* Explora los datos que contiene, explica brevemente qué variables existen.
* Quedarse únicamente con las columnas que consideres relevantes. Justifica la elección.

**Nota importante:**
Reduce la dimensión del dataset de forma aleatoria para obtener un dataset de sólo 200.000 registros. Todos los ejercicios deben realizarse con este dataset reducido.

**Las variables (columnas) estan definidas de la siguiente manera:**
* **Year**: 2008
* **Month**: 1-12
* **DayofMonth**: 1-31
* **DayOfWeek**: 1 (Monday) - 7 (Sunday)
* **DepTime**: actual departure time (local, hhmm)
* **CRSDepTime** scheduled departure time (local, hhmm)
* **ArrTime** actual arrival time (local, hhmm)
* **CRSArrTime** scheduled arrival time (local, hhmm)
* **UniqueCarrier** unique carrier code
* **FlightNum** flight number
* **TailNum** plane tail number: aircraft registration, unique aircraft identifier
* **ActualElapsedTime** in minutes
* **CRSElapsedTime** in minutes
* **AirTime** in minutes
* **ArrDelay arrival delay**, in minutes: A flight is counted as "on time" if it operated less than 15 minutes later the scheduled time shown in the carriers' Computerized Reservations Systems (CRS).
* **DepDelay** departure delay, in minutes
* **Origin** origin IATA airport code
* **Dest** destination IATA airport code
* **Distance** in miles
* **TaxiIn** taxi in time, in minutes
* **TaxiOut** taxi out time in minutes
* **Cancelled** *was the flight cancelled
* **CancellationCode** reason for cancellation (A = carrier, B = weather, C = NAS, D = security)
* **Diverted** 1 = yes, 0 = no
* **CarrierDelay** in minutes: Carrier delay is within the control of the air carrier. Examples of occurrences that may determine carrier delay are: aircraft cleaning, aircraft damage, awaiting the arrival of connecting passengers or crew, baggage, bird strike, cargo loading, catering, computer, outage-carrier equipment, crew legality (pilot or attendant rest), damage by hazardous goods, engineering inspection, fueling, handling disabled passengers, late crew, lavatory servicing, maintenance, oversales, potable water servicing, removal of unruly passenger, slow boarding or seating, stowing carry-on baggage, weight and balance delays.
* **WeatherDelay** in minutes: Weather delay is caused by extreme or hazardous weather conditions that are forecasted or manifest themselves on point of departure, enroute, or on point of arrival.
* **NASDelay** in minutes: Delay that is within the control of the National Airspace System (NAS) may include: non-extreme weather conditions, airport operations, heavy traffic volume, air traffic control, etc.
* **SecurityDelay** in minutes: Security delay is caused by evacuation of a terminal or concourse, re-boarding of aircraft because of security breach, inoperative screening equipment and/or long lines in excess of 29 minutes at screening areas.
* **LateAircraftDelay** in minutes: Arrival delay at an airport due to the late arrival of the same aircraft at a previous airport. The ripple effect of an earlier delay at downstream airports is referred to as delay propagation.

In [2]:
airline_data = pd.read_csv(r"C:/Users/Lenovo/Documents/Curso Barcelona Activa/IT academy/Data Science/IT_Academy_Spring_3Prog_num_DF_Stats/DelayedFlights.csv")

airline_data

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

### Análisis exploratorio del dataframe

In [3]:
airline_data.shape

(1936758, 30)

In [4]:
airline_data.columns

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

In [5]:
airline_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

**La data cuenta con 30 columnas con un total de registros de 1.936.757, con variables tipo float, int y objetos. Entre las 30 variables registradas se cuentan con 5 variable categoricas y 20 variables numericas. Las variables estan relacionadas en la salida y llegada de los vuelos, las causas, sus tiempos de retrasos (arribo y salida) planificados y reales, cancelacion de vuelos, desvios, en funcion de las aeolineas, lugar de origen y destino. Un punto importante de mencionar que Un vuelo se cuenta como "a tiempo" si se realizó con menos de 15 minutos de retraso con respecto a la hora programada que se muestra en los Sistemas de Reservas Computarizados (CRS) de las aerolíneas.**

* **Verificar registros duplicados:** 
Previo a la selección de las columnas a analizar y los 200.000 registros, se verificó si existían registros duplicados, y así descartar alguna desviación o que se solape algún resultados. 

In [6]:
airline_data.loc[airline_data.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, Year, Month, DayofMonth, DayOfWeek, DepTime, CRSDepTime, ArrTime, CRSArrTime, UniqueCarrier, FlightNum, TailNum, ActualElapsedTime, CRSElapsedTime, AirTime, ArrDelay, DepDelay, Origin, Dest, Distance, TaxiIn, TaxiOut, Cancelled, CancellationCode, Diverted, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay]
Index: []

[0 rows x 30 columns]

* **Verificar el número registros en la Columna 'Year'(año):** 
Para el inicio del descarte de algunas columnas, se decidió corrobarar que la data set estuviese comprendida por sólo un año, este caso el año 2008, utilizando la función nunique().

In [7]:
airline_data.Year.nunique()

1

* **Verificar las columnas Cancelled y Diverted:** 

In [8]:
airline_data['Cancelled'].mean()*100 #Para calcular el % 

0.032683484462178544

In [9]:
airline_data['Diverted'].mean()*100 #Para calcular el % 

0.40035977649246834

**Conclusiones:**
* No se encotraron registros duplicados en la data set.
* Se descartará la columna variable "Year", debido que se presenta como un valor unico, y coincide con la información de la fuente  https://www.kaggle.com/datasets/giovamata/airlinedelaycauses. 
* En las causas de los registros no indican algún retraso por el tipo de avión o engine, por lo que se descartará la variable "TailNum". 
* El análisis no se realizará en funcion de la hora de salida o llegada del vuelo, por lo que las variables 'DepTime', 'CRSDepTime', 'ArrTime' y 'CRSArrTime' se estarán descartando. 
* Como se evaluarán los tiempos y las causas de los retrasos de los vuelos, no se tomará en cuenta las cancelacion de vuelos, codigos de cancelación y desvio de vuelo, debido que el % de incidencia en la dataset es mínima de 0.03% y 0.4% respectivamente. 

* **Selección de las columnas a descartar:** De acuerdo a lo indicado en el punto anterior se descaratarán las columnas: **'Unnamed: 0', 'Year', 'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 'TailNum', 'Cancelled', 'CancellationCode', 'Diverted'.**

In [10]:
airline_data.drop(columns=['Unnamed: 0', 'Year', 'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 'TailNum', 'Cancelled', 'CancellationCode', 'Diverted'], inplace = True)
airline_data

Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  \
0            1           3          4            WN        335   
1            1           3          4            WN       3231   
2            1           3          4            WN        448   
3            1           3          4            WN       3920   
4            1           3          4            WN        378   
...        ...         ...        ...           ...        ...   
1936753     12          13          6            DL       1621   
1936754     12          13          6            DL       1631   
1936755     12          13          6            DL       1631   
1936756     12          13          6            DL       1639   
1936757     12          13          6            DL       1641   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
0                    128.0           150.0    116.0     -14.0       8.0   
1                    128.0           145.0    113.0       2.0      19.0   
2                     96.0            90.0     76.0      14.0       8.0   
3                     90.0            90.0     77.0      34.0      34.0   
4                    101.0           115.0     87.0      11.0      25.0   
...                    ...             ...      ...       ...       ...   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   
1936756              115.0           117.0     89.0       9.0      11.0   
1936757              123.0           135.0    104.0      -5.0       7.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  \
0          IAD  TPA       810     4.0      8.0           NaN           NaN   
1          IAD  TPA       810     5.0     10.0           NaN           NaN   
2          IND  BWI       515     3.0     17.0           NaN           NaN   
3          IND  BWI       515     3.0     10.0           2.0           0.0   
4          IND  JAX       688     4.0     10.0           NaN           NaN   
...        ...  ...       ...     ...      ...           ...           ...   
1936753    MSP  ATL       906     9.0     18.0           3.0           0.0   
1936754    RIC  ATL       481    15.0     34.0           0.0          57.0   
1936755    ATL  IAH       689     8.0     32.0           1.0           0.0   
1936756    IAD  ATL       533    13.0     13.0           NaN           NaN   
1936757    SAT  ATL       874     8.0     11.0           NaN           NaN   

         NASDelay  SecurityDelay  LateAircraftDelay  
0             NaN            NaN                NaN  
1             NaN            NaN                NaN  
2             NaN            NaN                NaN  
3             0.0            0.0               32.0  
4             NaN            NaN                NaN  
...           ...            ...                ...  
1936753       0.0            0.0               22.0  
1936754      18.0            0.0                0.0  
1936755      19.0            0.0               79.0  
1936756       NaN            NaN                NaN  
1936757       NaN            NaN                NaN  

[1936758 rows x 20 columns]

* **Extracción de datos:** Se seleccionó de manera aleatoria los 200.000 registros, utilizando el método sample(), definiendo el n como el número de registros.

In [11]:
new_data2008 = airline_data.sample(n=200000).copy().reset_index()
new_data2008

index  Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  \
0         36250      1          18          5            XE       2058   
1       1859501     12          17          3            FL        871   
2       1390844      8          25          1            AA        813   
3        889928      6           8          7            WN        798   
4        925001      6          30          1            XE       2674   
...         ...    ...         ...        ...           ...        ...   
199995   222966      2          11          1            XE       2609   
199996   497314      3          30          7            FL        440   
199997  1265509      7          17          4            DL       1650   
199998   397847      3          21          5            WN         50   
199999    99165      1          26          6            DL       1538   

        ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  ...  Origin  \
0                   139.0           183.0    121.0      76.0  ...     MCI   
1                   158.0           151.0    139.0      27.0  ...     ATL   
2                   183.0           185.0    162.0      13.0  ...     AUS   
3                   265.0           260.0    243.0      20.0  ...     SEA   
4                    77.0            85.0     56.0       1.0  ...     IAH   
...                   ...             ...      ...       ...  ...     ...   
199995              163.0           150.0    133.0      41.0  ...     EWR   
199996              106.0           125.0     92.0      95.0  ...     ATL   
199997               74.0            75.0     52.0      18.0  ...     JAX   
199998               53.0            60.0     44.0      11.0  ...     HRL   
199999              148.0           178.0    135.0      -5.0  ...     MSP   

       Dest Distance  TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  \
0       EWR     1092     9.0      9.0           0.0           0.0       5.0   
1       MSP      906     5.0     14.0          20.0           0.0       7.0   
2       LAX     1242    11.0     10.0           NaN           NaN       NaN   
3       BNA     1977     5.0     17.0           4.0           0.0       5.0   
4       LIT      374     5.0     16.0           NaN           NaN       NaN   
...     ...      ...     ...      ...           ...           ...       ...   
199995  MKE      725     6.0     24.0           2.0           0.0      13.0   
199996  MKE      669     3.0     11.0          95.0           0.0       0.0   
199997  ATL      270    10.0     12.0          18.0           0.0       0.0   
199998  HOU      276     4.0      5.0           NaN           NaN       NaN   
199999  SLC      991     5.0      8.0           NaN           NaN       NaN   

        SecurityDelay  LateAircraftDelay  
0                 0.0               71.0  
1                 0.0                0.0  
2                 NaN                NaN  
3                 0.0               11.0  
4                 NaN                NaN  
...               ...                ...  
199995            0.0               26.0  
199996            0.0                0.0  
199997            0.0                0.0  
199998            NaN                NaN  
199999            NaN                NaN  

[200000 rows x 21 columns]

## Ejercicio 2

**Realiza un informe completo del dataset:**

* Resume estadísticamente el dataset y las columnas de interés. Haz un análisis estadístico de lo que consideres relevante.
* Encuentra cuántos datos faltantes hay por columna.
* Crea nuevas columnas (velocidad media del vuelo, si ha llegado tarde o no...).
* Haz una tabla de las aerolíneas con mayores retrasos acumulados.
* ¿Cuáles son los vuelos más largos? ¿Y los más atrasados? Busca las rutas más largas y las que acumulen más retrasos.
* Aporta lo que consideres relevante.

**Premisa para el análisis:**

* Tiempo de retraso > 15 min
* Tiempo ontime <= 15 min

En función del Sistemas de Reservas Computarizados (CRS) de las aerolíneas.

* **Resume estadísticamente el dataset y las columnas de interés. Haz un análisis estadístico de lo que consideres relevante.**

In [12]:
# Analisis estadisticos de las varaibles numericas
new_data2008.describe().T

count           mean            std   min       25%  \
index              200000.0  969773.593475  560082.871114  13.0  484297.5   
Month              200000.0       6.120820       3.488547   1.0       3.0   
DayofMonth         200000.0      15.726600       8.773328   1.0       8.0   
DayOfWeek          200000.0       3.984605       1.992724   1.0       2.0   
FlightNum          200000.0    2184.435235    1942.501247   1.0     614.0   
ActualElapsedTime  199144.0     133.099566      71.714448  16.0      80.0   
CRSElapsedTime     199985.0     134.078111      71.058949 -12.0      81.0   
AirTime            199144.0     108.093862      68.368938   5.0      58.0   
ArrDelay           199144.0      42.283343      56.584057 -67.0       9.0   
DepDelay           200000.0      43.265585      53.224677   6.0      12.0   
Distance           200000.0     764.461730     572.692799  11.0     338.0   
TaxiIn             199278.0       6.791939       5.197125   0.0       4.0   
TaxiOut            199943.0      18.227740      14.418481   0.0      10.0   
CarrierDelay       129064.0      19.155551      42.792168   0.0       0.0   
WeatherDelay       129064.0       3.752712      21.665010   0.0       0.0   
NASDelay           129064.0      15.079294      34.173109   0.0       0.0   
SecurityDelay      129064.0       0.084237       2.195126   0.0       0.0   
LateAircraftDelay  129064.0      25.218171      41.977145   0.0       0.0   

                        50%         75%        max  
index              970757.0  1455441.25  1936752.0  
Month                   6.0        9.00       12.0  
DayofMonth             16.0       23.00       31.0  
DayOfWeek               4.0        6.00        7.0  
FlightNum            1545.0     3421.00     9741.0  
ActualElapsedTime     116.0      165.00      682.0  
CRSElapsedTime        116.0      165.00      660.0  
AirTime                90.0      137.00      645.0  
ArrDelay               25.0       56.00     1357.0  
DepDelay               24.0       53.00     1355.0  
Distance              606.0      997.00     4962.0  
TaxiIn                  6.0        8.00      200.0  
TaxiOut                14.0       21.00      342.0  
CarrierDelay            2.0       21.00     1291.0  
WeatherDelay            0.0        0.00      868.0  
NASDelay                1.0       15.00     1357.0  
SecurityDelay           0.0        0.00      357.0  
LateAircraftDelay       8.0       33.00     1254.0

In [13]:
#Analisis estadisticos de las variables categoricas
new_data2008.describe(include = ['O']).T

count unique  top   freq
UniqueCarrier  200000     20   WN  39004
Origin         200000    296  ATL  13784
Dest           200000    299  ORD  11174

In [130]:
new_data2008_corr = new_data2008[['ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance']].dropna().corr()
new_data2008_corr

ActualElapsedTime  CRSElapsedTime   AirTime  ArrDelay  \
ActualElapsedTime           1.000000        0.971687  0.977059  0.067516   
CRSElapsedTime              0.971687        1.000000  0.986264 -0.015980   
AirTime                     0.977059        0.986264  1.000000 -0.000470   
ArrDelay                    0.067516       -0.015980 -0.000470  1.000000   
DepDelay                    0.020490        0.008099 -0.001175  0.953340   
Distance                    0.953372        0.981966  0.980343 -0.029925   

                   DepDelay  Distance  
ActualElapsedTime  0.020490  0.953372  
CRSElapsedTime     0.008099  0.981966  
AirTime           -0.001175  0.980343  
ArrDelay           0.953340 -0.029925  
DepDelay           1.000000 -0.006155  
Distance          -0.006155  1.000000

**Observaciones:**
* En el analisis numérico se puede observar que no se tomaron en cuenta todos los registros, esto permite inferir que existen valores nulos o NaN, lo que pudierá conllevar arrojar errores en el calculo de las nuevas variables o columnas. Se procederá a verificar la cantidad de datos Nulos o faltantes con la funcion isnull.
* En las variables **ArrDelay y DepDelay** : se observa un tiempo medio de retraso en la llegada y salida de 42.22 min y 43.23 respectivamente y con desviación estandar relativamente más alta que los valores medio, por lo que se puede inferir que los datos no son muy dispersos. El 75% de los vuelos tienen un retraso en la salida y llegada del 56 min aproximadamente.	
* En la variable **Distance:** Se observa que la distancia media de los vuelos es de 764 millas y que el 50% de vuelos tienen una distacia aproximada de 606 millas.
* En el análisis categóricos, con la data reducida, se tienen 3 varaibles categóricas, indicando que de los 200.000 registros, se tienen 20 aerolíneas, 298 origenes y destinos  para el análisis, de cuales la aerolinea, la ciudad origen y destino con mayor frecuencia de vuelos fueron WN, ATL y ORD respectivamente.
* Con respecto al análisis de variables que correlacionan entre sí, se puede observar que ArrDelay y ArrDelay tienen una alta correlación(0.95) entre si,  asi como AirTime y Distance con un indice de correlación de 0.98. 

* **Encuentra cuántos datos faltantes hay por columna.**

In [15]:
new_data2008.isnull().sum().sort_values(ascending = False)

LateAircraftDelay    71037
SecurityDelay        71037
NASDelay             71037
WeatherDelay         71037
CarrierDelay         71037
ActualElapsedTime      885
AirTime                885
ArrDelay               885
TaxiIn                 740
TaxiOut                 50
CRSElapsedTime          21
Distance                 0
index                    0
Dest                     0
Origin                   0
Month                    0
FlightNum                0
UniqueCarrier            0
DayOfWeek                0
DayofMonth               0
DepDelay                 0
dtype: int64

In [16]:
new_data2008[new_data2008['AirTime'].isnull()].reset_index()

level_0    index  Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  \
0        332   346024      2           3          7            AA       1677   
1        810  1135960      7           2          3            OO       6155   
2       1245  1794858     12           7          7            OO       2880   
3       1905   351769      2           1          5            AQ         41   
4       2057  1695637     11          30          7            FL        956   
..       ...      ...    ...         ...        ...           ...        ...   
880   199384  1909819     12          26          5            AA       2363   
881   199533  1247469      7          24          4            B6        931   
882   199624    50590      1          25          5            OH       5274   
883   199958  1661039     11          30          7            OO       5595   
884   199989  1672825     11           9          7            US        262   

     ActualElapsedTime  CRSElapsedTime  AirTime  ...  Origin  Dest Distance  \
0                  NaN           265.0      NaN  ...     ORD   SNA     1726   
1                  NaN           124.0      NaN  ...     OKC   ORD      693   
2                  NaN           145.0      NaN  ...     MKE   EWR      725   
3                  NaN            47.0      NaN  ...     ITO   HNL      216   
4                  NaN           148.0      NaN  ...     BWI   FLL      925   
..                 ...             ...      ...  ...     ...   ...      ...   
880                NaN           170.0      NaN  ...     ORD   HDN     1009   
881                NaN           315.0      NaN  ...     ORD   LGB     1735   
882                NaN           120.0      NaN  ...     RDU   JFK      426   
883                NaN            55.0      NaN  ...     SAN   LAX      109   
884                NaN           323.0      NaN  ...     ANC   PHX     2551   

    TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0      NaN     22.0           NaN           NaN       NaN            NaN   
1      NaN     11.0           NaN           NaN       NaN            NaN   
2      NaN      NaN           NaN           NaN       NaN            NaN   
3      NaN      7.0           NaN           NaN       NaN            NaN   
4      5.0     17.0           NaN           NaN       NaN            NaN   
..     ...      ...           ...           ...       ...            ...   
880    NaN     50.0           NaN           NaN       NaN            NaN   
881    NaN     22.0           NaN           NaN       NaN            NaN   
882    NaN     24.0           NaN           NaN       NaN            NaN   
883    NaN     12.0           NaN           NaN       NaN            NaN   
884   20.0     13.0           NaN           NaN       NaN            NaN   

     LateAircraftDelay  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  
..                 ...  
880                NaN  
881                NaN  
882                NaN  
883                NaN  
884                NaN  

[885 rows x 22 columns]

In [17]:
new_data2008[['ActualElapsedTime', 'AirTime', 'ArrDelay', 'CRSElapsedTime', 'TaxiIn', 'TaxiOut','CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].reset_index()

index  ActualElapsedTime  AirTime  ArrDelay  CRSElapsedTime  TaxiIn  \
0            0              237.0    210.0      23.0           255.0     8.0   
1            1               79.0     55.0      83.0            87.0     4.0   
2            2              204.0    177.0      14.0           206.0     4.0   
3            3              238.0    193.0      80.0           200.0    12.0   
4            4               72.0     35.0      92.0            65.0     7.0   
...        ...                ...      ...       ...             ...     ...   
199995  199995              114.0     88.0      10.0           120.0    12.0   
199996  199996              106.0     87.0       6.0           110.0     3.0   
199997  199997              140.0    128.0      76.0           130.0     4.0   
199998  199998               73.0     63.0      14.0            88.0     3.0   
199999  199999              140.0    124.0       7.0           146.0     8.0   

        TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0          19.0           0.0           0.0       0.0            0.0   
1          20.0           0.0           0.0       0.0            0.0   
2          23.0           NaN           NaN       NaN            NaN   
3          33.0           0.0           7.0      38.0            0.0   
4          30.0           0.0           0.0       7.0            0.0   
...         ...           ...           ...       ...            ...   
199995     14.0           NaN           NaN       NaN            NaN   
199996     16.0           NaN           NaN       NaN            NaN   
199997      8.0          66.0           0.0      10.0            0.0   
199998      7.0           NaN           NaN       NaN            NaN   
199999      8.0           NaN           NaN       NaN            NaN   

        LateAircraftDelay  
0                    23.0  
1                    83.0  
2                     NaN  
3                    35.0  
4                    85.0  
...                   ...  
199995                NaN  
199996                NaN  
199997                0.0  
199998                NaN  
199999                NaN  

[200000 rows x 12 columns]

**Observaciones:**
* Las variables: CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay presentan 71.037 registros con NaN, presentando la mayor cantidad con respecto el resto de las variables. ActualElapsedTime, AirTime y ArrDelay con 885 registros de NaN,  TaxiIn con 740 registros, y TaxiOut con 50 y CRSElapsedTime con 21 regitros. 
* Se puede observar donde el 'ArrDelay' es menor a 15 min, algunos registros de las variables: CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay presentan "NaN", seran sustituidos el valor con 0.0.
* Se observa que los registros con valor "cero" en la columnas de CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, presentan "NaN" en algunos registros de la variable 'ArrDelay', en tal sentido se estaran sustituyendo el valor con 0.0.
* Con respectos a las variables ActualElapsedTime, AirTime, TaxiIn, TaxiOut y CRSElapsedTime, como son variables que se encuentran en función del tiempo serán sustituidas por su media (mean).

In [16]:
#En las causas y el tiempo de retrasado de los arribos del avion, sustituir los valores NaN con el valor 0.0.
new_data2008['WeatherDelay'] = new_data2008['WeatherDelay'].fillna(0)
new_data2008['SecurityDelay'] = new_data2008['SecurityDelay'].fillna(0)
new_data2008['ArrDelay'] = new_data2008['ArrDelay'].fillna(0)
new_data2008['CarrierDelay'] = new_data2008['CarrierDelay'].fillna(0)
new_data2008['NASDelay'] = new_data2008['NASDelay'].fillna(0)
new_data2008['LateAircraftDelay'] = new_data2008['LateAircraftDelay'].fillna(0)

# En los valores faltantes que las variables se encuentran en función del tiempo, sustituir por la media.
ActualElapsedTime_mean = new_data2008['ActualElapsedTime'].mean()
new_data2008['ActualElapsedTime'] = new_data2008['ActualElapsedTime'].fillna(ActualElapsedTime_mean)
AirTime_mean = new_data2008['AirTime'].mean()
new_data2008['AirTime'] = new_data2008['AirTime'].fillna(AirTime_mean)
TaxiIn_mean = new_data2008['TaxiIn'].mean()
new_data2008['TaxiIn'] = new_data2008['TaxiIn'].fillna(TaxiIn_mean)
TaxiOut_mean = new_data2008['TaxiOut'].mean()
new_data2008['TaxiOut'] = new_data2008['TaxiOut'].fillna(TaxiOut_mean)
CRSElapsedTime_mean = new_data2008['CRSElapsedTime'].mean()
new_data2008['CRSElapsedTime'] = new_data2008['CRSElapsedTime'].fillna(CRSElapsedTime_mean)

new_data2008

index  Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  \
0         36250      1          18          5            XE       2058   
1       1859501     12          17          3            FL        871   
2       1390844      8          25          1            AA        813   
3        889928      6           8          7            WN        798   
4        925001      6          30          1            XE       2674   
...         ...    ...         ...        ...           ...        ...   
199995   222966      2          11          1            XE       2609   
199996   497314      3          30          7            FL        440   
199997  1265509      7          17          4            DL       1650   
199998   397847      3          21          5            WN         50   
199999    99165      1          26          6            DL       1538   

        ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  ...  Origin  \
0                   139.0           183.0    121.0      76.0  ...     MCI   
1                   158.0           151.0    139.0      27.0  ...     ATL   
2                   183.0           185.0    162.0      13.0  ...     AUS   
3                   265.0           260.0    243.0      20.0  ...     SEA   
4                    77.0            85.0     56.0       1.0  ...     IAH   
...                   ...             ...      ...       ...  ...     ...   
199995              163.0           150.0    133.0      41.0  ...     EWR   
199996              106.0           125.0     92.0      95.0  ...     ATL   
199997               74.0            75.0     52.0      18.0  ...     JAX   
199998               53.0            60.0     44.0      11.0  ...     HRL   
199999              148.0           178.0    135.0      -5.0  ...     MSP   

       Dest Distance  TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  \
0       EWR     1092     9.0      9.0           0.0           0.0       5.0   
1       MSP      906     5.0     14.0          20.0           0.0       7.0   
2       LAX     1242    11.0     10.0           0.0           0.0       0.0   
3       BNA     1977     5.0     17.0           4.0           0.0       5.0   
4       LIT      374     5.0     16.0           0.0           0.0       0.0   
...     ...      ...     ...      ...           ...           ...       ...   
199995  MKE      725     6.0     24.0           2.0           0.0      13.0   
199996  MKE      669     3.0     11.0          95.0           0.0       0.0   
199997  ATL      270    10.0     12.0          18.0           0.0       0.0   
199998  HOU      276     4.0      5.0           0.0           0.0       0.0   
199999  SLC      991     5.0      8.0           0.0           0.0       0.0   

        SecurityDelay  LateAircraftDelay  
0                 0.0               71.0  
1                 0.0                0.0  
2                 0.0                0.0  
3                 0.0               11.0  
4                 0.0                0.0  
...               ...                ...  
199995            0.0               26.0  
199996            0.0                0.0  
199997            0.0                0.0  
199998            0.0                0.0  
199999            0.0                0.0  

[200000 rows x 21 columns]

In [17]:
new_data2008.isnull().sum().sort_values(ascending = False)

index                0
Origin               0
SecurityDelay        0
NASDelay             0
WeatherDelay         0
CarrierDelay         0
TaxiOut              0
TaxiIn               0
Distance             0
Dest                 0
DepDelay             0
Month                0
ArrDelay             0
AirTime              0
CRSElapsedTime       0
ActualElapsedTime    0
FlightNum            0
UniqueCarrier        0
DayOfWeek            0
DayofMonth           0
LateAircraftDelay    0
dtype: int64

* **Crear nuevas columnas**
    * Velocidad media del vuelo. 
    * Si ha llegado tarde o no. 
    * Salido tarde o no.
    * Total del retraso.

In [18]:
#Calculo de la velocidad media con las variables (columnas) 'Distance' y 'AirTime'.

new_data2008['flight_speed'] = new_data2008['Distance'] / (new_data2008['AirTime']/60)
new_data2008.head().reset_index()

level_0    index  Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  \
0        0    36250      1          18          5            XE       2058   
1        1  1859501     12          17          3            FL        871   
2        2  1390844      8          25          1            AA        813   
3        3   889928      6           8          7            WN        798   
4        4   925001      6          30          1            XE       2674   

   ActualElapsedTime  CRSElapsedTime  AirTime  ...  Dest  Distance TaxiIn  \
0              139.0           183.0    121.0  ...   EWR      1092    9.0   
1              158.0           151.0    139.0  ...   MSP       906    5.0   
2              183.0           185.0    162.0  ...   LAX      1242   11.0   
3              265.0           260.0    243.0  ...   BNA      1977    5.0   
4               77.0            85.0     56.0  ...   LIT       374    5.0   

  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0     9.0           0.0           0.0       5.0            0.0   
1    14.0          20.0           0.0       7.0            0.0   
2    10.0           0.0           0.0       0.0            0.0   
3    17.0           4.0           0.0       5.0            0.0   
4    16.0           0.0           0.0       0.0            0.0   

   LateAircraftDelay  flight_speed  
0               71.0    541.487603  
1                0.0    391.079137  
2                0.0    460.000000  
3               11.0    488.148148  
4                0.0    400.714286  

[5 rows x 23 columns]

In [19]:
#Identificar si esta retrasado en salida o llegada

def delay_interval(x): 
    if x <= 15: 
        mensaje = "Ontime"
    else: 
        mensaje = "Delay"
    return mensaje

new_data2008['Is_ArrDelay'] = new_data2008['ArrDelay'].apply(delay_interval)
                            
new_data2008['Is_DepDelay'] = new_data2008['DepDelay'].apply(delay_interval)

new_data2008.head().reset_index()

level_0    index  Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  \
0        0    36250      1          18          5            XE       2058   
1        1  1859501     12          17          3            FL        871   
2        2  1390844      8          25          1            AA        813   
3        3   889928      6           8          7            WN        798   
4        4   925001      6          30          1            XE       2674   

   ActualElapsedTime  CRSElapsedTime  AirTime  ...  TaxiIn  TaxiOut  \
0              139.0           183.0    121.0  ...     9.0      9.0   
1              158.0           151.0    139.0  ...     5.0     14.0   
2              183.0           185.0    162.0  ...    11.0     10.0   
3              265.0           260.0    243.0  ...     5.0     17.0   
4               77.0            85.0     56.0  ...     5.0     16.0   

  CarrierDelay WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  \
0          0.0          0.0       5.0            0.0               71.0   
1         20.0          0.0       7.0            0.0                0.0   
2          0.0          0.0       0.0            0.0                0.0   
3          4.0          0.0       5.0            0.0               11.0   
4          0.0          0.0       0.0            0.0                0.0   

   flight_speed  Is_ArrDelay  Is_DepDelay  
0    541.487603        Delay        Delay  
1    391.079137        Delay        Delay  
2    460.000000       Ontime       Ontime  
3    488.148148        Delay       Ontime  
4    400.714286       Ontime       Ontime  

[5 rows x 25 columns]

In [20]:
# Suma total de los retrasos (salida + llegada)
new_data2008["Total_Delay"] = new_data2008["ArrDelay"] + new_data2008["DepDelay"]
new_data = new_data2008
new_data

index  Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  \
0         36250      1          18          5            XE       2058   
1       1859501     12          17          3            FL        871   
2       1390844      8          25          1            AA        813   
3        889928      6           8          7            WN        798   
4        925001      6          30          1            XE       2674   
...         ...    ...         ...        ...           ...        ...   
199995   222966      2          11          1            XE       2609   
199996   497314      3          30          7            FL        440   
199997  1265509      7          17          4            DL       1650   
199998   397847      3          21          5            WN         50   
199999    99165      1          26          6            DL       1538   

        ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  ...  TaxiOut  \
0                   139.0           183.0    121.0      76.0  ...      9.0   
1                   158.0           151.0    139.0      27.0  ...     14.0   
2                   183.0           185.0    162.0      13.0  ...     10.0   
3                   265.0           260.0    243.0      20.0  ...     17.0   
4                    77.0            85.0     56.0       1.0  ...     16.0   
...                   ...             ...      ...       ...  ...      ...   
199995              163.0           150.0    133.0      41.0  ...     24.0   
199996              106.0           125.0     92.0      95.0  ...     11.0   
199997               74.0            75.0     52.0      18.0  ...     12.0   
199998               53.0            60.0     44.0      11.0  ...      5.0   
199999              148.0           178.0    135.0      -5.0  ...      8.0   

       CarrierDelay WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  \
0               0.0          0.0       5.0            0.0               71.0   
1              20.0          0.0       7.0            0.0                0.0   
2               0.0          0.0       0.0            0.0                0.0   
3               4.0          0.0       5.0            0.0               11.0   
4               0.0          0.0       0.0            0.0                0.0   
...             ...          ...       ...            ...                ...   
199995          2.0          0.0      13.0            0.0               26.0   
199996         95.0          0.0       0.0            0.0                0.0   
199997         18.0          0.0       0.0            0.0                0.0   
199998          0.0          0.0       0.0            0.0                0.0   
199999          0.0          0.0       0.0            0.0                0.0   

        flight_speed  Is_ArrDelay  Is_DepDelay  Total_Delay  
0         541.487603        Delay        Delay        196.0  
1         391.079137        Delay        Delay         47.0  
2         460.000000       Ontime       Ontime         28.0  
3         488.148148        Delay       Ontime         35.0  
4         400.714286       Ontime       Ontime         10.0  
...              ...          ...          ...          ...  
199995    327.067669        Delay        Delay         69.0  
199996    436.304348        Delay        Delay        209.0  
199997    311.538462        Delay        Delay         37.0  
199998    376.363636       Ontime        Delay         29.0  
199999    440.444444       Ontime        Delay         20.0  

[200000 rows x 25 columns]

* **Haz una tabla de las aerolíneas con mayores retrasos acumulados.**

In [31]:
new_data.groupby(["UniqueCarrier"])[["ArrDelay", "DepDelay","Total_Delay"]].agg(sum).sort_values(by="Total_Delay",ascending=False)

ArrDelay   DepDelay  Total_Delay
UniqueCarrier                                   
WN             1178424.0  1362385.0    2540809.0
AA              923027.0   927300.0    1850327.0
UA              701788.0   737515.0    1439303.0
MQ              647705.0   634689.0    1282394.0
OO              611628.0   609506.0    1221134.0
XE              534843.0   539154.0    1073997.0
DL              484242.0   472429.0     956671.0
CO              409909.0   440589.0     850498.0
EV              402840.0   409447.0     812287.0
US              370541.0   397572.0     768113.0
YV              381621.0   381819.0     763440.0
NW              368479.0   348241.0     716720.0
FL              315451.0   309786.0     625237.0
B6              307911.0   310400.0     618311.0
OH              270276.0   260265.0     530541.0
9E              251226.0   254105.0     505331.0
AS              140768.0   151270.0     292038.0
F9               81391.0    80418.0     161809.0
HA               22944.0    22481.0      45425.0
AQ                2364.0     2742.0       5106.0

La aerolínea con mayor retraso es la WN con 2.536.236 minutos

* **¿Cuáles son los vuelos más largos?**

In [23]:
new_data.groupby(['Origin', 'Dest'])[['Distance']].agg(max).sort_values(by="Distance",ascending=False)

Distance
Origin Dest          
EWR    HNL       4962
HNL    EWR       4962
       ATL       4502
ATL    HNL       4502
HNL    ORD       4243
...               ...
SNA    LAX         36
LAX    SNA         36
PSG    WRG         31
WRG    PSG         31
JFK    LGA         11

[4813 rows x 1 columns]

El vuelo más largo se realizó con en la ruta "EWR-HNL" con 4962 de millas de distancia y una velocidad media de 503.75 milla/hr.

* **¿Cuáles son los vuelos los más atrasados?**.

In [33]:
new_data.groupby(["UniqueCarrier","Origin", "Dest","Distance"])[["ArrDelay", "DepDelay","Total_Delay"]].agg(max).sort_values(by="Total_Delay",ascending=False)

ArrDelay  DepDelay  Total_Delay
UniqueCarrier Origin Dest Distance                                 
NW            HNL    MSP  3972        2461.0    2457.0       4918.0
              OMA    MSP  282         1583.0    1552.0       3135.0
              SMF    MSP  1518        1346.0    1351.0       2697.0
              GEG    MSP  1175        1289.0    1287.0       2576.0
AA            EGE    DFW  720         1292.0    1216.0       2508.0
...                                      ...       ...          ...
AQ            SNA    OGG  2513         -15.0      11.0         -8.0
EV            MCO    PFN  290          -17.0       8.0         -9.0
FL            TPA    MSP  1307         -16.0       7.0         -9.0
UA            LIH    DEN  3414         -19.0       9.0        -10.0
XE            CLE    OKC  948          -19.0       6.0        -13.0

[7599 rows x 3 columns]

El vuelo con mayor retraso fue por la aerolinea "NW" en la ruta "HNL-MSP" con una distancia de 3972 millas y tiempo de retraso de 4918 minutos

* **Vuelos por meses.**

In [34]:
new_data['Month'].value_counts()

12    21235
3     20856
6     20748
2     19302
7     18950
1     18891
8     16673
4     15968
5     15968
11    10824
10    10751
9      9834
Name: Month, dtype: int64

* **El valor medio de retraso en los vuelos por meses**

In [35]:
new_data.groupby('Month')['Total_Delay'].mean()

Month
1     86.362977
2     90.881929
3     85.005178
4     78.487162
5     76.323146
6     90.824561
7     93.318311
8     85.004558
9     72.577588
10    65.772951
11    78.531781
12    98.326065
Name: Total_Delay, dtype: float64

* **Números de vuelos a tiempo y retrasados en su llegada por meses**

In [36]:
new_data.groupby(['Month',"Is_ArrDelay"])[["FlightNum"]].count()

FlightNum
Month Is_ArrDelay           
1     Delay            12211
      Ontime            6680
2     Delay            12882
      Ontime            6420
3     Delay            13265
      Ontime            7591
4     Delay             9785
      Ontime            6183
5     Delay             9557
      Ontime            6411
6     Delay            13663
      Ontime            7085
7     Delay            12163
      Ontime            6787
8     Delay             9987
      Ontime            6686
9     Delay             5569
      Ontime            4265
10    Delay             5713
      Ontime            5038
11    Delay             6361
      Ontime            4463
12    Delay            14466
      Ontime            6769

* **Tiempo acumulado de los vuelos por retrasos y a tiempos en su llegadas por meses**

In [37]:
new_data.groupby(['Month',"Is_ArrDelay"])[["ArrDelay"]].sum()

ArrDelay
Month Is_ArrDelay           
1     Delay         781765.0
      Ontime         25991.0
2     Delay         847605.0
      Ontime         28091.0
3     Delay         840832.0
      Ontime         32154.0
4     Delay         592110.0
      Ontime         25791.0
5     Delay         569235.0
      Ontime         26568.0
6     Delay         911516.0
      Ontime         33295.0
7     Delay         847768.0
      Ontime         29035.0
8     Delay         660880.0
      Ontime         26534.0
9     Delay         325060.0
      Ontime         15639.0
10    Delay         316865.0
      Ontime         17176.0
11    Delay         397755.0
      Ontime         13542.0
12    Delay        1015374.0
      Ontime         26797.0

* **Tiempo total por causas de retrasos en función de los meses en la llegada de los vuelos**

In [38]:
causes= new_data.groupby(['Month'])['CarrierDelay','WeatherDelay','NASDelay','SecurityDelay','LateAircraftDelay'].sum()
causes

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_19560\1484378073.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  causes= new_data.groupby(['Month'])['CarrierDelay','WeatherDelay','NASDelay','SecurityDelay','LateAircraftDelay'].sum()


CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay
Month                                                                        
1          237673.0       44711.0  183259.0         1334.0           319498.0
2          243643.0       58838.0  188843.0         1187.0           360059.0
3          267771.0       46572.0  178319.0         1322.0           351873.0
4          198675.0       26394.0  135176.0          917.0           235028.0
5          178635.0       25239.0  137096.0          912.0           231208.0
6          252797.0       57484.0  241845.0         1069.0           363451.0
7          249312.0       56722.0  221167.0          822.0           324065.0
8          220640.0       40740.0  149650.0          800.0           253280.0
9          124330.0       13618.0   80035.0          297.0           109540.0
10         118689.0       10301.0   71602.0          448.0           118960.0
11         116766.0       14324.0  110814.0          273.0           158053.0
12         277917.0       76613.0  221456.0         1296.0           443522.0

* **Vuelos por dias**

In [39]:
new_data['DayOfWeek'].value_counts()

5    33353
1    30240
4    30057
7    29514
3    27102
2    27013
6    22721
Name: DayOfWeek, dtype: int64

* **Números de vuelos a tiempo y retrasados en función del dia.**

In [40]:
new_data.groupby(['DayOfWeek',"Is_ArrDelay"])[["FlightNum"]].count()

FlightNum
DayOfWeek Is_ArrDelay           
1         Delay            19068
          Ontime           11172
2         Delay            17178
          Ontime            9835
3         Delay            16772
          Ontime           10330
4         Delay            18765
          Ontime           11292
5         Delay            21587
          Ontime           11766
6         Delay            13597
          Ontime            9124
7         Delay            18655
          Ontime           10859

**Conclusiones:**
* Se observa un tiempo medio de retraso en la llegada y salida de 42.22 min y 43.23 respectivamente. 
* La aerolínea con mayor retraso es la "WN" con 2.536.236 minutos.
* El vuelo más largo se realizó en la ruta "EWR-HNL" con 4962 de millas de distancia.
* El vuelo con mayor retraso fue por la aerolínea "NW" en la ruta "HNL-MSP" con una distancia de 3972 millas y tiempo de retraso de 4918 minutos.
* Los meses con mayor retraso son los meses de invierno Diciembre, Enero, Febrero y Marzo.
* Los dias con mayor retraso son Lunes y Viernes.
* La causa de retraso en la llegada con mayor tiempo es LateAircraftDelay, esto es debido al efecto domino entre la salida y llegada de los vuelos, lo cual se puede corroborar, con coeficiente de correlacion de casi 1 (0.95) entre las dos variables.

## Exporta el dataset limpio y con las nuevas columnas a Excel.

In [28]:
new_data.to_excel('new_data2008.xlsx', index = False)